beetsplug/edit.py

```diff
@@ -313,8 +352,9 @@ def apply_data(self, objs, old_data, new_data):
             if forbidden:
                 continue
 
-            id = int(old_dict['id'])
-            apply(obj_by_id[id], new_dict)
+            # Reconcile back the user edits, using the reference_field.
+            val = self.obj_from_ref(old_dict)
```

Comment:
Continuing the theme, I'm also in general somewhat saddened by the need for all this extra complexity to identify objects by an appropriate field.

Here's one other crazy idea: give the objects a *temporary id* for the purpose of this workflow. Then, after the process is done, remove it before continuing with the import process.